tkhq/qos
# refactor build intermediates into lfs for dev/ci

## Goal

Ensure both local dev and CI pull a pre-built and compressed toolchain container image instead of continually rebuilding from scratch. Dependencies of this toolchain image will -not- need to be pulled unless the toolchain image itself is being rebuilt, due to a willful change of the contents of CONFIG_DIR.

## In Scope

Address the following problems currently faced by engineers:

- Those who clone with GIT_LFS_SKIP_SMUDGE are in a broken state unable to use "make" commands at all
- Engineers are sometimes unable to reproduce builds in a branch as they are using a toolchain image built from their previous branch
- Engineers/CI must to pull down multiple GB of dependencies of toolchain.tar in order to build it
- toolchain.tar image is over 1.5gb and expensive to remotely cache and restore
- Simplify Github actions to take advantage of built-in LFS based toolchain caching
  - We should strive to ensure that CI is always testing the same flow people would use locally
- Engineers/CA must manually rebuild expensive and rarely changed dependencies like the linux kernel and rust-libstd

## Out of scope

- ```make reproduce``` will still skip all caching and build everything from 0
- Github LFS is still very slow, capping out at around 10MB/s.
  - In the future toolchain.tar could be in our own much faster ECR as a primary and LFS as the backup.
  - Still 1m to pull a prebuilt image is much faster than pulling gigs of deps and building toolchain from 0

## Changes

- [x] Ensure toolchain submodule is cloned automatically on a fresh clone
- [x] The ~1.5GB toolchain.tar becomes a ~500MB toolchain.tgz
- [x] The toolchain.tgz file is built/tagged against the hash of the ```config``` dir
- [x] Engineers who clone with GIT_LFS_SKIP_SMUDGE will have the toolchain.tgz pulled from LFS on demand
- [x] If the toolchain.tgz file is out of date due to changes in ```config``` dir, needed dependencies will be fetched only then 
- [x] linux kernel cached in LFS
- [x] musl rust libstd cached in LFS
- [x] static openssl cached in LFS
- [x] nsm.ko cached in LFS
- [x] github actions transfers caches from LFS to internal cache for cost savings
- [ ] dist artifacts are forced to the timestamp of release.env
  - deals with edge cases where code is newer but results in identical dist hashes 

## Testing

1. Clone repo with LFS disabled
    ```
    export GIT_LFS_SKIP_SMUDGE=true
    git clone --single-branch --branch lrvick/cache-toolchain git@github.com:tkhq/qos qos-ss
    cd qos-ss
    ```
2. Build tree using cached dist and toolchain

    ```
    make
    ```
3. Confirm the following behavior
   - cache/toolhain.tgz is downloaded from LFS
   - dist/* artifacts are downloaded from LFS
   - dist and toolchain are -not- rebuilt and make is mostly a NOOP
